src/Microsoft.AspNet.Mvc.Core/Internal/Routing/LinkGenerationDecisionTree.cs

```diff
@@ -131,19 +133,36 @@ private class LinkGenerationMatchComparer : IComparer<LinkGenerationMatch>
 
             public int Compare(LinkGenerationMatch x, LinkGenerationMatch y)
             {
-                // For these comparisons lower is better.
-
+                //For this comparison lower is better
                 if (x.Entry.Order != y.Entry.Order)
                 {
                     return x.Entry.Order.CompareTo(y.Entry.Order);
                 }
 
+                if(x.ParameterCount != y.ParameterCount)
+                {
+                    if(x.Entry?.Template.Segments?.Count != y.Entry?.Template.Segments?.Count)
```

Comment:
Imagine we have these routes:
help/general
help
help/{id}

The reason for this PR is that when someone generates a url from urlhelper.Action pointing at the Action help with a parameter Id we want to get "help/1234" not "help?id=1234" (that is, it should match the third route, not the second). But we also want to have the more concrete match try to match first. So if someone goes looking for  "help/general" it needs to match the first route, not the third.